### DISCUSSION TITLE
Prioritize results by most used (frecency)

### DISCUSSION THREAD

**Discussion by Gremious** [Flow-Launcher/Flow.Launcher#2496] [Ideas]
[Frecency](https://en.wikipedia.org/wiki/Frecency) is a simple concept, used mainly by browsers, of calculating some priority based on frequent + recent usage of an item.

I am wondering if it would be possible for (ideally) flow or at least a plugin (doubtful) to implement this for search results.

The concept is fairly simple. I have a fresh install, I search "steam", I get this:

![image](https://github.com/Flow-Launcher/Flow.Launcher/assets/21365711/c4dea377-0709-4fa4-9a21-2cfb0f961bab)

I then go and actually select 'msteams'.

Following that, msteams should show up first when I search for steam because it has been accessed more than steam (1 vs 0 in this example) and more recently too.

The same should apply for plugin results too, e.g. I have a plugin that generates discord timestamps:

![image](https://github.com/Flow-Launcher/Flow.Launcher/assets/21365711/1850d96c-4aef-45c0-a800-1f841da7a63c)

But I only ever use the "x hours from now" format. So after using it a few times, it would be ideal for it to be the top result.

As a side-note, Listary implements something like this and it has been _very_ nice

**Comment by Garulf** [comment]
Under hood this is already happening. Each time you select a result its score is bumped ever so slightly higher next time you search that query. 

IIRC this does not transfer between different queries with similar results.

**Reply by Gremious** [reply]
Ah yeah, I think I see it if I try a couple times.

As you say, it seems to be specifically per query rather than per item. Per item weights would definitely be nicer if possible imo, especially for plugins. As it stands I would have to write e.g. `dt 8pm` for every hour and use that said hour a couple times until it figures out what I prefer.

**Reply by taooceros** [reply]
We also have per item. It's just the bumping is smaller than per query.

**Reply by Ishmaeel** [reply]
Per item bump should be huge. I launch a program, chances are I'll launch it again soon.
Is there a setting where we can fiddle with the bump value? I may be willing to recompile from source.

**Reply by Garulf** [reply]
> Per item bump should be huge. I launch a program, chances are I'll launch it again soon.
> 
> Is there a setting where we can fiddle with the bump value? I may be willing to recompile from source.

There is a "Pin to top" feature within the context menu for every item.

**Reply by Ishmaeel** [reply]
Yes, but that is also per-query, not per-item.

**Comment by taooceros** [comment]
For recency we would be happy to accept pr. But I cannot think of a good model for that atm.

**Comment by spartanhooah** [comment]
This doesn't seem to work if I have to type additional letters to get my target to even show up on the list. For example, if I want "PatchMyPC", I have to type "patchm", and PatchMyPC is the first on the list. There doesn't seem to be any way to get it to show up with a shorter query string, though. Anything I can do differently to adjust that?

**Reply by jjw24** [reply]
Try 'pat' and scroll down the list (yes, unfortunately) to find PatchMyPC, go to context menus and pin to top, then trigger it a couple times. Once the score is high enough you shouldn't need to pin it anymore.

**Comment by manusgeek** [comment]
+1 one this. It would be great if FlowLauncher could prioritise the most access programs with the first couple of letters typed, and show him on top. A bit like launchy or chrome search bar. FlowLauncher is not doing that very well currently.

**Reply by deefrawley** [reply]
But it does do this now. I use Pot Player as my media player. At first I had to type `pot p` and find it in the list. Soon just typing `pot` surfaced it to the top. Now I simply type `p` and it is the first result in the list. The same thing has happened for all of the common apps I launch through Flow so I'm not sure what you are seeing, or if you haven't given it enough time but I feel this is working exactly as intended.

**Reply by manusgeek** [reply]
I don't know why it works in your case and not mine. I use version 1.19.5. 
The most frequent program i use is "editlastscreenshot.bat". When type everything up to "edit", it doesn't show in the list at all.
After I type "editl" it shows up in second position below a google search for this string. Even after i've typed "editlastscreenshot.bat" fully, the google search for this (which i never click on) is still the first result. I've used and clicked on editlastscreenshot.bat around 100 times, if I include all the testing i did around this. I also tried alternating between Everything and "Windows Index" in Explorer plugins config but that didn't make a difference.

**Reply by Ishmaeel** [reply]
The issue is closely related to Explorer plugin. When that one is activated, the initial search results are saturated with hundreds of files and this list is truncated before getting ranked, so your desired result might not even appear with a short search term. When it does appear, the tiny, tiny recency score is not enough to push it up the list unless you launch it dozens of times to make it accumulate a significant score.

I have the same issue and I resolved it in two ways: In my personal computer I use a custom version I compiled where I bump the recency bonus to an obscene number. In my work computer, I just disabled the Explorer plugin and only kept Programs.

**Reply by manusgeek** [reply]
Thanks, that's really useful to understand this, super helpful. Do you know why google search remains the first result while the other result(that i want) stays in second position. Even when "editlastscreenshot.bat" is fully typed?

**Reply by Ishmaeel** [reply]
That, I wouldn't know, I have the web search keyword disabled, but if I had to guess, I'd say it's because of plugin priority.
